### PR TITLE
chore: Plural data sources in TPF leveraging on resource data source

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Plural data sources in TPF leveraging on resource data source.

Checks to make sure current plural data source schemas are generated the same as the auto-generated ones are removed in this commit: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2844/commits/1f2c2e5e1bc1a3d96967e09599b7c5bffacc447c

Link to any related issue(s): CLOUDP-287359

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals I have added appropriate changelog entries.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: docs/data-sources/flex_clusters.md [hunk 1]
```diff
@@ -39,9 +39,7 @@ output "mongodbatlas_flex_clusters_names" {
 
 ### Required
 
-- `project_id` (String) Unique 24-hexadecimal digit string that identifies your project. Use the [/groups](#tag/Projects/operation/listProjects) endpoint to retrieve all projects to which the authenticated user has access.
-
-**NOTE**: Groups and projects are synonymous terms. Your group id is the same as your project id. For existing groups, your group/project id remains the same. The resource and corresponding endpoints use the term groups.
+- `project_id` (String) Unique 24-hexadecimal character string that identifies the project.
```
Comment (lantoli) [inline docs/data-sources/flex_clusters.md:42]: updates so description in plural ds is the same as in singular ds

File: internal/service/advancedclustertpf/data_source.go [hunk 2]
```diff
@@ -32,3 +31,9 @@ func (d *ds) Schema(ctx context.Context, req datasource.SchemaRequest, resp *dat
 
 func (d *ds) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
 }
+
+// TODO: see if resource model can be used instead, probably different only in timeouts
+type ModelDS struct {
```
Comment (lantoli) [inline internal/service/advancedclustertpf/data_source.go:37]: move model to data_source as schema is now auto-generated for singular and plural ds so schema files don't have the schema any more

File: internal/service/advancedclustertpf/plural_data_source_schema.go [hunk 3]
```diff
@@ -1,57 +0,0 @@
...
-
-import (
-	"context"
-
-	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
-	"github.com/hashicorp/terraform-plugin-framework/types"
-)
-
-// TODO: see if we can leverage on resource or singular data source schema, e.g. have a func to add computed
-func PluralDataSourceSchema(ctx context.Context) schema.Schema {
```
Comment (lantoli) [inline internal/service/advancedclustertpf/plural_data_source_schema.go:11]: plural ds schema is now auto-generated

File: docs/data-sources/encryption_at_rest_private_endpoints.md [hunk 4]
```diff
@@ -25,12 +25,12 @@ output "number_of_endpoints" {
...
 ### Required
 
-- `cloud_provider` (String) Label that identifies the cloud provider of the private endpoint.
+- `cloud_provider` (String) Label that identifies the cloud provider for the Encryption At Rest private endpoint.
 - `project_id` (String) Unique 24-hexadecimal digit string that identifies your project.
 
 ### Read-Only
 
-- `results` (Attributes List) List of returned documents that MongoDB Cloud providers when completing this request. (see [below for nested schema](#nestedatt--results))
+- `results` (Attributes List) List of returned documents that MongoDB Cloud provides when completing this request. (see [below for nested schema](#nestedatt--results))
```
Comment (corryroot) [inline docs/data-sources/encryption_at_rest_private_endpoints.md]:
```suggestion
- `results` (Attributes List) List of documents that MongoDB Cloud returns for this request. To learn more, see the following [nested schema](#nestedatt--results).
```

File: docs/data-sources/resource_policies.md [hunk 5]
```diff
@@ -98,7 +98,7 @@ output "policy_ids" {
 ### Read-Only
 
 - `resource_policies` (Attributes List, Deprecated) (see [below for nested schema](#nestedatt--resource_policies))
-- `results` (Attributes List) (see [below for nested schema](#nestedatt--results))
+- `results` (Attributes List) List of returned documents that MongoDB Cloud provides when completing this request. (see [below for nested schema](#nestedatt--results))
```
Comment (corryroot) [inline docs/data-sources/resource_policies.md]:
```suggestion
- `results` (Attributes List) List of documents that MongoDB Cloud returns for this request. To learn more, see the following [nested schema](#nestedatt--results).
```

File: internal/common/conversion/schema_generation.go [hunk 6]
```diff
@@ -11,14 +11,44 @@ import (
...
+	overrideFields(attrs, overridenFields)
+	rootAttrs := convertAttrs(rs.Attributes, requiredFields)
+	for name := range rootAttrs {
+		if !slices.Contains(requiredFields, name) {
+			delete(rootAttrs, name)
 		}
 	}
-	ds := dsschema.Schema{Attributes: attrs, Blocks: blocks}
+	overrideFields(rootAttrs, overridenRootFields)
+	resultsDoc := "List of returned documents that MongoDB Cloud provides when completing this request."
```
Comment (corryroot) [inline internal/common/conversion/schema_generation.go]:
```suggestion
	resultsDoc := "List of documents that MongoDB Cloud returns for this request."
```

File: internal/common/conversion/schema_generation_test.go [hunk 7]
```diff
@@ -264,6 +264,194 @@ func TestDataSourceSchemaFromResource(t *testing.T) {
...
+							Description:         "desc overridenString",
+							MarkdownDescription: "desc overridenString",
+							Validators: []validator.String{
+								stringvalidator.ConflictsWith(path.MatchRoot("otherAttr")),
+							},
+						},
+					},
+				},
+				Description:         "List of returned documents that MongoDB Cloud provides when completing this request.",
+				MarkdownDescription: "List of returned documents that MongoDB Cloud provides when completing this request.",
```
Comment (corryroot) [inline internal/common/conversion/schema_generation_test.go]:
```suggestion
				Description:         "List of documents that MongoDB Cloud returns for this request.",
				MarkdownDescription: List of documents that MongoDB Cloud returns for this request.",
```

File: internal/common/conversion/schema_generation_test.go [hunk 8]
```diff
@@ -264,6 +264,194 @@ func TestDataSourceSchemaFromResource(t *testing.T) {
...
+						},
+						"attrString": dsschema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "desc attrString",
+							Description:         "desc attrString",
+						},
+					},
+				},
+				Description:         "List of returned documents that MongoDB Cloud provides when completing this request.",
+				MarkdownDescription: "List of returned documents that MongoDB Cloud provides when completing this request.",
```
Comment (corryroot) [inline internal/common/conversion/schema_generation_test.go]:
```suggestion
				Description:         "List of documents that MongoDB Cloud returns for this request.",
				MarkdownDescription: "List of documents that MongoDB Cloud returns for this request.",
```

**Review by corryroot — APPROVED** [review]
LGTM! I commented with some copy nits.

File: internal/common/conversion/schema_generation.go [hunk 9]
```diff
@@ -11,14 +11,44 @@ import (
...
-			delete(attrs, name)
-		} else {
-			attrs[name] = attr
+	overrideFields(attrs, overridenFields)
+	ds := dsschema.Schema{Attributes: attrs, Blocks: blocks}
+	UpdateSchemaDescription(&ds)
+	return ds
+}
+
+func PluralDataSourceSchemaFromResource(rs schema.Schema, requiredFields []string, overridenFields, overridenRootFields map[string]dsschema.Attribute, overrideResultsDoc string, hasLegacyFields bool) dsschema.Schema {
```
Comment (EspenAlbert) [inline internal/common/conversion/schema_generation.go]: Seems like a lot of arguments, can we use a Config struct instead?

File: internal/common/conversion/schema_generation.go [hunk 10]
```diff
@@ -11,14 +11,44 @@ import (
...
+	resultsDoc := "List of returned documents that MongoDB Cloud provides when completing this request."
+	if overrideResultsDoc != "" {
+		resultsDoc = overrideResultsDoc
+	}
+	rootAttrs["results"] = dsschema.ListNestedAttribute{
+		Computed: true,
+		NestedObject: dsschema.NestedAttributeObject{
+			Attributes: attrs,
+		},
+		MarkdownDescription: resultsDoc,
```
Comment (EspenAlbert) [inline internal/common/conversion/schema_generation.go:56]: I guess there is no way we can use a better description here?
Or generic is ok?

File: internal/common/conversion/schema_generation.go [hunk 11]
```diff
@@ -11,14 +11,44 @@ import (
...
+	}
+	if hasLegacyFields {
+		rootAttrs["id"] = dsschema.StringAttribute{Computed: true}
```
Comment (EspenAlbert) [inline internal/common/conversion/schema_generation.go:59]: why is this necessary?

**Review by EspenAlbert — APPROVED** [review]
Great! Like how well this is going.
A few minor comments.

File: internal/common/conversion/schema_generation.go (hunk 11 above)
Comment (lantoli) [reply internal/common/conversion/schema_generation.go:59]: it's needed for streamconnection and streaminstance, and maybe for more resources if we do the migration to TPF

File: internal/common/conversion/schema_generation.go (hunk 10 above)
Comment (lantoli) [reply internal/common/conversion/schema_generation.go:56]: do you mean to improve the var name resultsDoc, or the text "List of documents that MongoDB ..." ?

Comment (EspenAlbert) [reply internal/common/conversion/schema_generation.go:56]: "List of documents that MongoDB"

File: internal/common/conversion/schema_generation.go (hunk 11 above)
Comment (EspenAlbert) [reply internal/common/conversion/schema_generation.go:59]: I think this was a limitation of SDKv2, however, expecting new resources to not wanting this low level pagination args. So, ok to keep the `id`

**Review by oarbusi — APPROVED** [review]
LGTM

File: internal/common/conversion/schema_generation.go (hunk 10 above)
Comment (lantoli) [reply internal/common/conversion/schema_generation.go:56]: this is the one we're using right now in most of the resources, do you have a suggestion?

(specific resources can override the generic one with param overrideResultsDoc)

**Comment by lantoli** [comment]
thanks @corryroot , suggestions applied in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2844/commits/32c95030ca49f066a3796044feb22b382f8a8756

I can't change the last part as it's auto-generated.

File: internal/common/conversion/schema_generation.go (hunk 11 above)
Comment (lantoli) [reply internal/common/conversion/schema_generation.go:59]: yeah, i added "legacy" in the param name to make it clear that we should try to avoid using it

File: internal/common/conversion/schema_generation.go (hunk 9 above)
Comment (lantoli) [reply internal/common/conversion/schema_generation.go]: you're totally right, changed in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2844/commits/28ab07f8624eb4bc20c1b4dcf7129d7266d31245

also this will allow adding more params without changing all resources every time

File: internal/common/conversion/schema_generation.go (hunk 10 above)
Comment (EspenAlbert) [reply internal/common/conversion/schema_generation.go:56]: Aah, perfect, `overrideResultsDoc` should do the job 👍
